dai: nxp: sai: Add support for all synchronization modes

LaurentiuM1234 opened the pull request:
Previously, we only supported RX in SYNC with TX. Since the SAI can also support ASYNC RX with ASYNC TX and ASYNC RX with SYNC TX, this patch series adds supports for these scenarios as well.

Also, due to the i.MX93 errata `ERR051421`, we need to add support for the case in which one of the directions is in SYNC with the other while the BYP bit is set for the ASYNC direction.

Below you may find some waveforms proving that BCLK uses the ASYNC direction's divided BCLK instead of the bypassed one when BYP is toggled.

DIV is 0 => RX_BCLK is half of MCLK (should be equal to MCLK or TX_BCLK)
![Screenshot from 2023-12-19 13-07-32](https://github.com/zephyrproject-rtos/zephyr/assets/26043217/7d4662e9-698d-439e-9e67-49fba7d9ae62)

(BCLK is half of MCLK's frequency, which is wrong. See commit description and comments from the code for details)

DIV is 1 => RX_BCLK is MCLK / 4 (should be equal to MCLK or TX_BCLK)
![Screenshot from 2023-12-19 13-09-51](https://github.com/zephyrproject-rtos/zephyr/assets/26043217/79ca9693-4ad7-4d2f-918d-e409a82fe922)

Note: MCLK is approximately 12.5Mhz in both scenarios.

**EDIT**: Also attaching some register dumps for:

1) First scenario
![Screenshot from 2023-12-19 11-33-30](https://github.com/zephyrproject-rtos/zephyr/assets/26043217/18a8222e-3661-4f2e-aff8-801c0a5ab22a)

2) Second scenario
![Screenshot from 2023-12-19 11-37-39](https://github.com/zephyrproject-rtos/zephyr/assets/26043217/0df27c59-f8da-4560-bfa7-47f8f11825fa)

dbaluta commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -353,6 +353,68 @@ static int sai_config_set(const struct device *dev,
+
+#ifdef CONFIG_SAI_IMX93_ERRATA_051421
+	/* a few notes about the following piece of code:
+	 *
```
This looks very good. What do you think about adding a function in order to improve readability of sai_config_set?

e.g 
```
#ifdef CONFIG_SAI_IMX93_ERRATA_051421

void sai_config_set_errata_05152(struct *device...)

{ 
//code for errata
}

#else

void void sai_config_set_errata_05152(struct *device...)
{
///
}

#endif

```

LaurentiuM1234 commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -528,14 +555,89 @@ static int sai_trigger_stop(const struct device *dev,
 {
 	struct sai_data *data;
+	const struct sai_config *cfg;
 	uint32_t old_state;
```
Do you mean something like:

Commit 1: introduce DTS properties and assign them to the `tx_config`/`rx_config` variables in `sai_config_set()`
Commit 2: introduce additional logic for the various sync modes

The issue with this is that if you stop at commit 1 and don't take commit 2 you'll end up with something broken (i.e: you can only use RX sync with TX even if your properties allow you to also configure TX sync with RX and TX, RX async)

dbaluta replied:
Exactly, with the additional quirk that in sai_config_set() will not read tx_config/rx_config properties from dts but will set the default values.

And then in the second patch we will actually use the values from dts.

But if you find that this still breaks the series we can drop the idea.

LaurentiuM1234 replied:
Ah, got it. Will give it a shot. Should work fine.

LaurentiuM1234 committed:
drivers: dai: sai: Introduce the rx_sync_mode/tx_sync_mode properties

In preparation for supporting all synchronization modes, this
commit introduces the rx_sync_mode/tx_sync_mode DTS properties.
Using these, the user will be able to specify which synchronization
mode the SAI should use.

At the moment, the driver does nothing with the values from
said properties but still checks if their values are sane
(i.e: it checks if the directions are both in SYNC mode which
is forbidden). By default, if "rx_sync_mode" or "tx_sync_mode"
is not specified, the direction will be set to ASYNC mode.
As such, below one may find a couple of valid examples
depicting this idea:

	tx_sync_mode = <0>;
	rx_sync_mode = <0>;

is the same as not specifying any of the properties,

	tx_sync_mode = <1>;
	rx_sync_mode = <0>;

is the same as:

	tx_sync_mode = <1>;

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
drivers: dai: sai: Add support for all SYNC modes

With the introduction of the "rx_sync_mode" and "tx_sync_mode"
properties, the user may choose which synchronization mode the
SAI should use. To support this, the code had to be changed a
bit such that the software reset and the disable operations
work on all synchronization modes.

What this commit does, is it changes the software reset and
disable sequences such that they work with any of the
supported synchronization modes. Also, the syncMode field
of sai_transceiver_t structure is set to the value passed
from the DTS during sai_config_set().

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
drivers: dai: sai: Add fix for i.MX93 SAI errata 051421

ERRATA 051421 states that if the SAI is FSYNC/BCLK master,
one of the directions is SYNC with the other, and the
ASYNC direction has the BYP bit toggled, the SYNC direction's
BCLK won't be generated properly.

This commit fixes this issue by enabling BCI for the SYNC
direction. What this does is it makes the SYNC direction use
the BCLK from the ASYNC direction's pad, which, if the BYP
bit is toggled, will be the undivided MCLK. Without this fix,
the SYNC direction will use the ASYNC direction's BCLK obtained
by dividing the MCLK via the following formula: MCLK / ((DIV + 1) * 2).
This is wrong because the ASYNC direction will use an undivided
MCLK while the SYNC direction will use a divided MCLK.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented:
**V2 changes**:

1. Initial commit was split into 2: 1 introducing `rx_sync_mode` and `tx_sync_mode` properties and setting their associated fields in `sai_config`, 2nd introducing the additional logic.
2. Errata logic has been moved to a separate function alongside the wall of text explaining the errata. Should be more readable now.
3. Wall of text from inside `sai_rx_tx_sw_reset()` has been moved to the beginning of the function. Should improve readability as well.

dbaluta commented:
Looks good to me. Thanks @LaurentiuM1234 . Ready to merge.

iuliana-prodan commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -383,8 +381,33 @@ static int sai_config_set(const struct device *dev,
+	 * hardware states and their software states. The software
+	 * states can be obtained by reading data->tx/rx_enabled, while
+	 * the hardware states can be obtained by reading TCSR/RCSR. The
+	 * hadrware state can actually differ from the software state.
```
s/hadrware /hardware

iuliana-prodan commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -528,14 +553,77 @@ static int sai_trigger_stop(const struct device *dev,
+ *
+ *	Combination d) is not valid and is covered by a BUILD_ASSERT(). As such, there's 3 valid
+ *	combintations that need to be supported. Since the main branch of the IF statement covers
+ *	combination a), there's only combinatons b) and c) to be covered here.
```
s/ combinatons/combinations

iuliana-prodan commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -142,6 +142,73 @@ static const struct dai_properties
+ *
+ *	IMPORTANT2: for this to work, you also need to enable the
+ *	pad's input path. For i.MX93, this can be achieved by setting
+ *	the pad's SION bit.
```
As it is now, this works fine on i.MX93 I suppose, right?
The SION bit is set somewhere else? Where?

LaurentiuM1234 replied:
Yep, works fine.
As for the SION bit: it's set via an IOMUXC register (e.g: `SW_MUX_CTL_PAD_GPIO_IO18`). We don't need to set it via this driver as it's already handled by Linux in our setup. One thing to watch out for here is the fact that the `cs42448` DTS for 93 doesn't set this bit so one would have to modify it to support the SYNC-ASYNC scenario (this is why the default configuration is ASNYC-ASYNC).

iuliana-prodan reviewed (commented):
LGTM.. just some typos

LaurentiuM1234 commented:
**V3 changes**

- Fixed typos pointed out by @iuliana-prodan. No functional change.

fabiobaltieri commented:
(nxp specific, switching assignee)
